File: training/models/latent_interface.py
"""ELIT-style latent interface: Read -> latent processing -> Write.

Decouples spatial token count from core compute. K learned latent tokens
compress the spatial tokens, process them through transformer blocks,
then write updated information back to the spatial tokens.

Supports tail dropping during training for importance ordering.
"""

from __future__ import annotations

import torch
import torch.nn as nn


class LatentInterface(nn.Module):
    """ELIT-style latent interface: Read -> latent processing -> Write.

    Decouples spatial token count from core compute. K learned latent tokens
    compress the 20 spatial tokens, process them through transformer blocks,
    then write updated information back to the spatial tokens.

    Supports tail dropping during training for importance ordering.
    """

    def __init__(self, d_model: int = 128, n_latents: int = 12, n_heads: int = 8,
                 n_latent_blocks: int = 2):
        super().__init__()
        self.d_model = d_model
        self.n_latents = n_latents

        # Learned latent tokens
        self.latents = nn.Parameter(torch.zeros(n_latents, d_model))
        nn.init.normal_(self.latents, std=0.02)

        # Read: latents attend to spatial tokens (cross-attention)
        self.read_attn = nn.MultiheadAttention(d_model, n_heads, dropout=0.0, batch_first=True)
        self.read_norm_q = nn.LayerNorm(d_model)
        self.read_norm_kv = nn.LayerNorm(d_model)
        self.read_ff = nn.Sequential(
            nn.Linear(d_model, d_model * 2), nn.GELU(), nn.Linear(d_model * 2, d_model),
        )
        self.read_ff_norm = nn.LayerNorm(d_model)

        # Latent transformer blocks (self-attention among latents)
        self.latent_blocks = nn.ModuleList()
        for _ in range(n_latent_blocks):
            self.latent_blocks.append(nn.TransformerEncoderLayer(
                d_model=d_model, nhead=n_heads, dim_feedforward=d_model * 2,
                dropout=0.0, activation="gelu", batch_first=True, norm_first=True,
            ))

        # Write: spatial tokens attend to latents (cross-attention)
        self.write_attn = nn.MultiheadAttention(d_model, n_heads, dropout=0.0, batch_first=True)
        self.write_norm_q = nn.LayerNorm(d_model)
        self.write_norm_kv = nn.LayerNorm(d_model)
        self.write_ff = nn.Sequential(
            nn.Linear(d_model, d_model * 2), nn.GELU(), nn.Linear(d_model * 2, d_model),
        )
        self.write_ff_norm = nn.LayerNorm(d_model)

        # Zero-init Write output projection -> identity pass-through at init
        nn.init.zeros_(self.write_attn.out_proj.weight)
        nn.init.zeros_(self.write_attn.out_proj.bias)

    def forward(
        self,
        spatial_tokens: torch.Tensor,
        spatial_mask: torch.Tensor | None = None,
        n_latents_override: int | None = None,
    ) -> tuple[torch.Tensor, torch.Tensor]:
        """
        Args:
            spatial_tokens: (B, N, d) -- entity/threat/position/aggregate tokens
            spatial_mask: (B, N) -- True where padded
            n_latents_override: if set, use only first J latents (tail dropping)
        Returns:
            updated_tokens: (B, N, d) -- spatial tokens enriched by latent processing
            pooled_latents: (B, d) -- mean of latent tokens (for temporal cell + heads)
        """
        B = spatial_tokens.shape[0]
        K = n_latents_override if n_latents_override is not None else self.n_latents

        # Initialize latents -- take first K (importance ordering via tail dropping)
        L = self.latents[:K].unsqueeze(0).expand(B, -1, -1)  # (B, K, d)

        # Read: latents attend to spatial tokens
        q = self.read_norm_q(L)
        kv = self.read_norm_kv(spatial_tokens)
        read_out, _ = self.read_attn(q, kv, kv, key_padding_mask=spatial_mask)
        L = L + read_out
        L = L + self.read_ff(self.read_ff_norm(L))

        # Latent self-attention blocks
        for block in self.latent_blocks:
            L = block(L)

        # Write: spatial tokens attend to latents
        q_w = self.write_norm_q(spatial_tokens)
        kv_w = self.write_norm_kv(L)
        write_out, _ = self.write_attn(q_w, kv_w, kv_w)
        updated_tokens = spatial_tokens + write_out
        updated_tokens = updated_tokens + self.write_ff(self.write_ff_norm(updated_tokens))

        # Pool latents for downstream heads
        pooled_latents = L.mean(dim=1)  # (B, d)

        return updated_tokens, pooled_latents

File: src/ai/core/burn_model/latent_interface.rs
//! ELIT-style Latent Interface for V6.
//!
//! K learned latent tokens compress the entity token sequence via Read/Write
//! cross-attention steps with latent self-attention in between.
//!
//! - Read: latents attend to (spatially-enriched) entity tokens
//! - 2x latent transformer blocks (self-attention)
//! - Write: entity tokens updated by latents (zero-init output proj at init)
//!
//! Tail dropping: during training, randomly use a subset of latents.
//! At inference, pick J based on compute budget.

use burn::module::{Module, Param};
use burn::nn::{
    Gelu, LayerNorm, LayerNormConfig, Linear, LinearConfig,
};
use burn::nn::transformer::{
    TransformerEncoder, TransformerEncoderConfig, TransformerEncoderInput,
};
use burn::prelude::*;
use burn::tensor::activation::softmax;

use super::config::*;

#[derive(Module, Debug)]
pub struct LatentInterface<B: Backend> {
    /// Learned latent tokens: [K, d_model]
    latents: Param<Tensor<B, 2>>,
    /// Read: latents (Q) attend to entity tokens (K/V)
    read_q: Linear<B>,
    read_k: Linear<B>,
    read_v: Linear<B>,
    read_out: Linear<B>,
    read_norm: LayerNorm<B>,
    /// Latent self-attention blocks
    latent_blocks: TransformerEncoder<B>,
    /// Write: entity tokens (Q) attend to latents (K/V)
    write_q: Linear<B>,
    write_k: Linear<B>,
    write_v: Linear<B>,
    write_out: Linear<B>,
    write_norm: LayerNorm<B>,
    n_heads: usize,
    d_model: usize,
    n_latents: usize,
    scale: f32,
}

#[derive(Config, Debug)]
pub struct LatentInterfaceConfig {
    #[config(default = "D_MODEL")]
    pub d_model: usize,
    #[config(default = "N_HEADS")]
    pub n_heads: usize,
    #[config(default = "N_LATENT_TOKENS")]
    pub n_latents: usize,
    #[config(default = "N_LATENT_BLOCKS")]
    pub n_latent_blocks: usize,
}

impl LatentInterfaceConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> LatentInterface<B> {
        let d = self.d_model;
        let head_dim = d / self.n_heads;

        // Initialize latents with small normal
        let latents = Tensor::random(
            [self.n_latents, d],
            burn::tensor::Distribution::Normal(0.0, 0.02),
            device,
        );

        // Read cross-attention
        let read_q = LinearConfig::new(d, d).init(device);
        let read_k = LinearConfig::new(d, d).init(device);
        let read_v = LinearConfig::new(d, d).init(device);
        let read_out = LinearConfig::new(d, d).init(device);
        let read_norm = LayerNormConfig::new(d).init(device);

        // Latent self-attention blocks
        let latent_blocks = TransformerEncoderConfig::new(d, d * 2, self.n_heads, self.n_latent_blocks)
            .with_norm_first(true)
            .with_dropout(0.0)
            .init(device);

        // Write cross-attention — zero-init output proj for identity passthrough
        let write_q = LinearConfig::new(d, d).init(device);
        let write_k = LinearConfig::new(d, d).init(device);
        let write_v = LinearConfig::new(d, d).init(device);
        let mut write_out = LinearConfig::new(d, d).init(device);
        // Zero-init write output projection: at init, write step is identity
        write_out.weight = Param::from_tensor(Tensor::zeros([d, d], device));
        write_out.bias = write_out.bias.map(|_| Param::from_tensor(Tensor::zeros([d], device)));

        let write_norm = LayerNormConfig::new(d).init(device);

        LatentInterface {
            latents: Param::from_tensor(latents),
            read_q, read_k, read_v, read_out, read_norm,
            latent_blocks,
            write_q, write_k, write_v, write_out, write_norm,
            n_heads: self.n_heads,
            d_model: d,
            n_latents: self.n_latents,
            scale: (head_dim as f32).sqrt().recip(),
        }
    }
}

impl<B: Backend> LatentInterface<B> {
    pub fn n_latents(&self) -> usize {
        self.n_latents
    }

    /// Forward pass through the latent interface.
    ///
    /// entity_tokens: [B, S, d] — output of entity encoder (+ spatial enrichment)
    /// entity_mask: [B, S] — true = valid token
    /// n_latents_override: if Some, use this many latents (for inference budget control)
    ///
    /// Returns: (updated_entity_tokens [B, S, d], pooled_latents [B, d])
    pub fn forward(
        &self,
        entity_tokens: Tensor<B, 3>,
        entity_mask: Tensor<B, 2, Bool>,
        n_latents_override: Option<usize>,
    ) -> (Tensor<B, 3>, Tensor<B, 2>) {
        let [batch, seq_len, _] = entity_tokens.dims();
        let d = self.d_model;
        let h = self.n_heads;
        let dh = d / h;

        // Determine number of latents to use
        let n_lat = n_latents_override.unwrap_or(self.n_latents);
        let n_lat = n_lat.min(self.n_latents);

        // Expand latents: [K, d] -> [B, n_lat, d]
        let latents = self.latents.val()
            .slice([0..n_lat])
            .unsqueeze_dim::<3>(0)
            .expand([batch, n_lat, d]);

        // --- Read step: latents attend to entity tokens ---
        let read_q: Tensor<B, 4> = self.read_q.forward(latents.clone())
            .reshape([batch, n_lat, h, dh]).swap_dims(1, 2); // [B, H, K, dh]
        let read_k: Tensor<B, 4> = self.read_k.forward(entity_tokens.clone())
            .reshape([batch, seq_len, h, dh]).swap_dims(1, 2); // [B, H, S, dh]
        let read_v: Tensor<B, 4> = self.read_v.forward(entity_tokens.clone())
            .reshape([batch, seq_len, h, dh]).swap_dims(1, 2); // [B, H, S, dh]

        let read_scores = read_q.matmul(read_k.swap_dims(2, 3)) * self.scale; // [B, H, K, S]

        // Mask padded entity tokens
        let pad_mask = entity_mask.clone().bool_not()
            .reshape([batch, 1, 1, seq_len])
            .expand([batch, h, n_lat, seq_len]);
        let read_scores = read_scores.mask_fill(pad_mask, -1e9);

        let read_weights = softmax(read_scores, 3);
        let read_out = read_weights.matmul(read_v); // [B, H, K, dh]
        let read_out: Tensor<B, 3> = read_out.swap_dims(1, 2).reshape([batch, n_lat, d]);
        let read_out = self.read_out.forward(read_out);

        // Residual + norm
        let latents = self.read_norm.forward(latents + read_out);

        // --- Latent self-attention blocks ---
        let input = TransformerEncoderInput::new(latents);
        let latents = self.latent_blocks.forward(input);

        // --- Write step: entity tokens attend to latents ---
        let write_q: Tensor<B, 4> = self.write_q.forward(entity_tokens.clone())
            .reshape([batch, seq_len, h, dh]).swap_dims(1, 2); // [B, H, S, dh]
        let write_k: Tensor<B, 4> = self.write_k.forward(latents.clone())
            .reshape([batch, n_lat, h, dh]).swap_dims(1, 2); // [B, H, K, dh]
        let write_v: Tensor<B, 4> = self.write_v.forward(latents.clone())
            .reshape([batch, n_lat, h, dh]).swap_dims(1, 2); // [B, H, K, dh]

        let write_scores = write_q.matmul(write_k.swap_dims(2, 3)) * self.scale; // [B, H, S, K]
        // No masking needed — all latents are valid
        let write_weights = softmax(write_scores, 3);
        let write_out = write_weights.matmul(write_v); // [B, H, S, dh]
        let write_out: Tensor<B, 3> = write_out.swap_dims(1, 2).reshape([batch, seq_len, d]);
        let write_out = self.write_out.forward(write_out);

        // Residual + norm (zero-init write_out means this starts as identity)
        let entity_tokens = self.write_norm.forward(entity_tokens + write_out);

        // Pool latents for temporal cell input
        let pooled: Tensor<B, 2> = latents.mean_dim(1).squeeze_dim::<2>(1); // [B, d]

        (entity_tokens, pooled)
    }
}
